### ISSUE TITLE
Failed to connect using Docker

### ISSUE THREAD

**Issue by Nickmiste** [Cysharp/MagicOnion#675]
I am trying to dockerize my MagicOnion server, and I keep getting the following error on the client when trying to call a service method:

```
Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="failed to connect to all addresses", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1693935927.809079000","description":"Failed to pick subchannel","file":"/tmpfs/altsrc/github/grpc/workspace_csharp_ext_macos_x64/src/core/ext/filters/client_channel/client_channel.cc","file_line":3218,"referenced_errors":[{"created":"@1693935927.809079000","description":"failed to connect to all addresses","file":"/tmpfs/altsrc/github/grpc/workspace_csharp_ext_macos_x64/src/core/lib/transport/error_utils.cc","file_line":165,"grpc_status":14}]}")
```

Everything was working when running directly on my host machine, so this issue has something to do with Docker. A few things to note:

- Client is Unity
- I am using an apple silicon mac (but Intel Unity)
- Using HTTP (connecting to `http://localhost:5000`)
- Server logs are identical when running in or out of docker

Here is my dockerfile:
```Dockerfile
# Build Image
FROM mcr.microsoft.com/dotnet/sdk:7.0-bookworm-slim AS build
WORKDIR /tmp/build

COPY . .

RUN dotnet publish -o /app

# App Image
FROM mcr.microsoft.com/dotnet/sdk:7.0-bookworm-slim
WORKDIR /app

COPY --from=build /app .

EXPOSE 5000

CMD ["dotnet", "Server.dll"]
```

and my docker-compose.yml:
```yml
version: "3.9"

services:
  magiconion:
    image: magiconion
    build:
      context: .
      dockerfile: Dockerfile
    container_name: magiconion
    ports:
      - 5000:5000
```

Has anyone run into this?

**Comment by mwegner** [comment]
This is a bit out of scope for MagicOnion itself, but port 5000 is the default _development_ port.  Very likely this container is listening on port 80.  Two options:

* Redirect port 5000 to 80 in your compose file (`- 5000:80` instead of `- 5000:80`)
* Add this line to your `Dockerfile`: `ENV ASPNETCORE_URLS=http://+:5000/`

**Comment by Nickmiste** [comment]
> This is a bit out of scope for MagicOnion itself, but port 5000 is the default _development_ port. Very likely this container is listening on port 80. Two options:
> 
> * Redirect port 5000 to 80 in your compose file (`- 5000:80` instead of `- 5000:5000`)
> * Add this line to your `Dockerfile`: `ENV ASPNETCORE_URLS=http://+:5000/`

No difference. I'm fairly certain that the server is listening on port 5000 beacuse of the container logs and that fact that I'm able to connect on port 5000 outside of docker.

Container logs:
```
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /app
````

**Comment by mwegner** [comment]
Oh huh!  I've definitely used ports inside Intel containers on macOS Apple Silicon machines (not with MagicOnion in particular, but I very much doubt that matters).

Maybe something else is already listening on port 5000 at the OS level?  I don't think the stuff inside the container will know.  You could also try explicitly using `127.0.0.1` instead of localhost, and binding to it in the docker compose definition with: `- 127.0.0.1:5000:5000`

(And connect to `127.0.0.1` instead of `localhost` resolution)

**Comment by Nickmiste** [comment]
Issue resolved. The fix was to have the server listen on `0.0.0.0` instead of `localhost`.

**Comment by mwegner** [comment]
Ah, right!  I vaguely remember there being some weirdness around localhost and Docker (the container's "localhost" isn't the same as the host's "localhost" or something along those lines).
